### PR TITLE
Working Agreement: GitHub Teams and Access

### PR DESCRIPTION
**Pull request by ameowlia**
* Everyone is welcome to review and comment.
* We will discuss this in the TOC meeting today (Nov 9th). 
* In the TOC meeting next week the TOC members will vote on whether to accept this working agreement.

### CODE REVIEW

**Review by LeePorte — APPROVED** [review]

**Review by stephanme — APPROVED** [review]

File: toc/working-agreements/2021-11-9-github-teams.md
Comment (emalm) [inline toc/working-agreements/2021-11-9-github-teams.md:1]:
```suggestion
# GitHub Teams and Access
```

File: toc/working-agreements/2021-11-9-github-teams.md
Comment (emalm) [inline toc/working-agreements/2021-11-9-github-teams.md:11]:
```suggestion
## Problem
CFF technical governance has changed to be organized around working groups and new roles like Tech Lead and Approver. Most working groups also have sub-groups.

We need ways within the community to:
* Provide the team members with proper push access to certain repos,
* Provide a way for others to tag groups of people.

## Solution

We will make GitHub teams within the `cloudfoundry` org to reflect these groups.

```

File: toc/working-agreements/2021-11-9-github-teams.md
Comment (emalm) [inline toc/working-agreements/2021-11-9-github-teams.md:19]:
```suggestion
| Name of Team  | Team Membership  | Permissions  |
|---|---|---|
| toc | Technical Oversight Committee | Admin access to everything |
| wg-leads | Tech and Execution Leads for all WGs | Write access to community repo |
| wg-[WORKING-GROUP-NAME] | All approvers and leads for a WG | None: only for organization and tagging |
| wg-[WORKING-GROUP-NAME]-leads | All leads for a WG | Admin access for all WG repos |
| wg-[WORKING-GROUP-NAME]-[SUBGROUP-NAME]-approvers | Approvers for a subgroup within a WG | Write access for all repos in the subgroup |
| wg-[WORKING-GROUP-NAME]-bots | Bot accounts for a WG | Write access for all WG repos |
```

* `[WORKING-GROUP-NAME]` is the name of a particular working group, converted to kebab case. 
* `[SUBGROUP-NAME]` is the name of a subgroup within a particular working group, converted to kebab case.

**Review by emalm — CHANGES_REQUESTED** [review]
Suggested some edits, but overall I approve once they are incorporated.

File: toc/working-agreements/2021-11-9-github-teams.md
Comment (rkoster) [reply toc/working-agreements/2021-11-9-github-teams.md:19]: the approvers group should also include the tech leads in my opinion. Because that will allow us to use [auto-assignment](https://docs.github.com/en/organizations/organizing-members-into-teams/managing-code-review-settings-for-your-team#about-auto-assignment).

Comment (loewenstein) [reply toc/working-agreements/2021-11-9-github-teams.md:19]: Could be done by including all -leads tesms into the corresponding -approvers teams.

Comment (rkoster) [reply toc/working-agreements/2021-11-9-github-teams.md:19]: Do approvers actually need write access? When using Code Review Limits approvers would actually only need explicit read access on the repo: 
Screenshot 2021-11-26 at 20 53 36

Just thinking about the least permissions required to make things work. We would still need something to merge the actual PRs but maybe that could be an auto-merge bot, once the required number of reviewer approvals has been reached.

Comment (loewenstein) [reply toc/working-agreements/2021-11-9-github-teams.md:19]: Isn't setting labels and assigning oneself to issue limited to write access as well? This could also be handled by by a bot reacting to comands though. Kubernetes and Istio communities have examples for that as well.

Comment (rkoster) [reply toc/working-agreements/2021-11-9-github-teams.md:19]: Having the wg and area names concatenated will result in really long names: `wg-foundational-infrastructure-system-logging-and-metrics-approvers`.

Does it make sense to just put area teams inside a parent team and not concatenate the names?

Comment (loewenstein) [reply toc/working-agreements/2021-11-9-github-teams.md:19]: If we can avoid name clashes by means other than namespacing, I don't see a problem here (if one can easily navigate from child to parent team).

Comment (totherme) [reply toc/working-agreements/2021-11-9-github-teams.md:19]: What permissions are needed on repos to set up CI integrations? I'm thinking both about "checks" that call out to external concourse/etc deployments, and also github actions for folks who might go that way.

**Comment by LeePorte** [comment]
@ameowlia There are a few thing on this PR that need resolving, can you have a look when you get a chance?

**Comment by ameowlia** [comment]
Wow, lots of comments; thank you for sharing all of your thoughts. The comments automatically got hidden when I merged @emalm 's suggestion (sorry!). 

1. @rkoster -  I added tech leads to the approvers group so they will be included in assignment. 
2. @loewenstein -  I like your idea of "including all -leads teams into the corresponding -approvers teams", but I will leave this as an implementation detail.
3. @rkoster - I'm inclined to let approvers have write access. This proposal is not enforcing that teams use bots to merge their PRs. For examples, reviewers may want to do this "by hand" in order to check their CI and make sure it is green before moving forward.
4. @rkoster + @loewenstein - Yes the names are long. @rkoster , I'm not quite sure what you mean by "Does it make sense to just put area teams inside a parent team and not concatenate the names?". I'm inclined to just let the names be long. We won't be typing them very often. Plus a defined order means that the names can be deduced without looking them up.
5. @totherme  -  Re "What permissions are needed on repos to set up CI integrations?". My guess is: it depends! I don't know of any CI integrations my teams have set up with CI and I'm not sure what you are doing 👀 . I suspect auditing everything would be time consuming. Unless you are certain that this WILL break something, I am inclined to try this, see it break, and change the permissions as needed. I suspect that there will be a lot of little changes once we actually enact this proposal.

**Comment by emalm** [comment]
Hey, @ameowlia, I had a few other suggestions that I think got lost in the shuffle, so I've re-requested them.

**Comment by ameowlia** [comment]
Updated to be in the RFC format.

File: toc/rfc/rfc-draft-github-teams-and-access.md
Comment (emalm) [inline toc/rfc/rfc-draft-github-teams-and-access.md:21]:
```suggestion
* Provide a way for others to tag groups of people.
```

**Review by emalm — CHANGES_REQUESTED** [review]
Looks good to me! The only change I requested is to remove a stray part of the RFC template doc.

**Review by emalm — APPROVED** [review]

**Review by dsboulder — APPROVED** [review]

**Comment by rkoster** [comment]
When trying to generate the subgroup names based on the current working group charter files, I'm getting the following list:
```
wg-app-runtime-deployments-cf-deployment-approvers
wg-app-runtime-interfaces-autoscaler-approvers
wg-app-runtime-interfaces-buildpacks-approvers
wg-app-runtime-interfaces-capi-approvers
wg-app-runtime-interfaces-cli-approvers
wg-app-runtime-interfaces-docs-approvers
wg-app-runtime-interfaces-java-tools-approvers
wg-app-runtime-interfaces-multiapps-approvers
wg-app-runtime-interfaces-notifications-approvers
wg-app-runtime-platform-diego-approvers
wg-app-runtime-platform-docs-approvers
wg-app-runtime-platform-garden-containers-approvers
wg-app-runtime-platform-logging-and-metrics-approvers
wg-app-runtime-platform-networking-approvers
wg-cf-on-k8s-cf-on-k8s-approvers
wg-foundational-infrastructure-credential-management-(credhub)-approvers
wg-foundational-infrastructure-disaster-recovery-(bbr)-approvers
wg-foundational-infrastructure-identity-and-auth-(uaa)-approvers
wg-foundational-infrastructure-integrated-databases-approvers
wg-foundational-infrastructure-system-logging-and-metrics-approvers
wg-foundational-infrastructure-vm-deployment-lifecycle-(bosh)-approvers
wg-service-management-cloud-service-broker-approvers
wg-service-management-osbapi-approvers
wg-service-management-servicefabrik-approvers
wg-service-management-volume-service-adapters-approvers
```

The names containing parenthesis look a bit weird (eg. `wg-foundational-infrastructure-credential-management-(credhub)-approvers`). Maybe it would be better to giving working group areas a name and a title (eg. `{name: credhub, title: credential-managment}`.

File: toc/rfc/rfc-draft-github-teams-and-access.md
Comment (emalm) [inline toc/rfc/rfc-draft-github-teams-and-access.md:12]:
```suggestion
All working groups SHOULD create GitHub teams using the standardized name format.
```

File: toc/rfc/rfc-draft-github-teams-and-access.md
Comment (emalm) [inline toc/rfc/rfc-draft-github-teams-and-access.md:34]: I'd suggest we standardize on "area" as the name for a WG subgroup and clarify how to convert WG and area names into these team name segments:
```suggestion
| wg-[WORKING-GROUP-NAME]-[AREA-NAME]-approvers | Approvers & tech leads for an area within a WG | Write access for all repos in the area |
| wg-[WORKING-GROUP-NAME]-bots | Bot accounts for a WG | Write access for all WG repos |

Where: 
* `WORKING-GROUP-NAME` is the name of the Working Group, converted to kebab case,
* `AREA-NAME` is the name of the area, also converted to kebab case, or a suitable short name that identifies it clearly and uniquely within the Working Group.
```

**Review by emalm — CHANGES_REQUESTED** [review]
@ameowlia I'm suggesting some changes based on @rkoster's comment about generating team names programmatically from the specification in this RFC. PTAL!

**Review by emalm — APPROVED** [review]

**Review by loewenstein — APPROVED** [review]

**Pull request merged**
